### PR TITLE
Possible reentrancy issues with hooks in mutate closure

### PR DESCRIPTION
**Pull request by shamb0**
Analysis of the issue "Possible reentrancy issues with hooks in mutate closure"

[Analysis In-progress | Initial analysis report](https://github.com/paritytech/substrate/issues/7605)

### CODE REVIEW

**Review by xlc — COMMENTED** [review]
You need to have `mod tests_reentrancy` in the balances/src/lib.rs file to include the new tests

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -946,39 +946,101 @@ impl<T: Config<I>, I: Instance> Currency<T::AccountId> for Module<T, I> where
...
+				if to_acc.total() < T::ExistentialDeposit::get() {
+					*maybe_to_acc = None;
+				} else {
+					*maybe_to_acc = Some(to_acc.clone());
+				}
+
+				let maybe_from_acc_endowed = if from_acc_is_new { Some(from_acc.free) } else { None };
+				let maybe_to_acc_endowed = if to_acc_is_new { Some(to_acc.free) } else { None };
 
+				if let Some(from_acc_endowed) = maybe_from_acc_endowed {
```
Comment (xlc) [inline frame/balances/src/lib.rs:1010]: what is the point of having `maybe_from_acc_endowed`?

This should be the same?

```
if from_acc_is_new {
                    Self::deposit_event(
 						RawEvent::Endowed(transactor.clone(), from_acc.free)
 					);
}
```

File: frame/balances/src/lib.rs [hunk 2]
```diff
@@ -946,39 +946,101 @@ impl<T: Config<I>, I: Instance> Currency<T::AccountId> for Module<T, I> where
...
+					Self::deposit_event(
+						RawEvent::Endowed(dest.clone(), to_acc_endowed)
+					);
+				}
 				Ok(())
 			})
 		})?;
 
+		if maybe_from_acc_call_post_mutation {
+			let _ = Self::post_mutation(transactor, from_acc);
```
Comment (xlc) [inline frame/balances/src/lib.rs:1026]: I don't think `let _ =` is needed. It is always a code smell to ignore thing without any explanation

**Review by xlc — COMMENTED** [review]
If the tests are passing, I guess it works.. just the logic is very confusing... but not sure what else can be done other than micro optimizing the code flow

File: frame/balances/src/lib.rs (hunk 2 above)
Comment (shamb0) [reply frame/balances/src/lib.rs]: Sure will ignore it.

File: frame/balances/src/lib.rs (hunk 1 above)
Comment (shamb0) [reply frame/balances/src/lib.rs]: Indeed good one,  refactored the code as suggested.

**Comment by shamb0** [comment]
> You need to have `mod tests_reentrancy` in the balances/src/lib.rs file to include the new tests

Thanks @xlc for comments, Yes included "mod tests_reentrancy;" part of  "balances/src/lib.rs"

File: frame/balances/src/lib.rs [hunk 3]
```diff
@@ -946,39 +947,98 @@ impl<T: Config<I>, I: Instance> Currency<T::AccountId> for Module<T, I> where
...
+				to_acc.free = to_acc
+					.free.checked_add(&value)
+					.ok_or(Error::<T, I>::Overflow)?;
 
 				let ed = T::ExistentialDeposit::get();
-				ensure!(to_account.total() >= ed, Error::<T, I>::ExistentialDeposit);
+				ensure!(
+					to_acc.total() >= ed,
+					Error::<T, I>::ExistentialDeposit,
+				);
```
Comment (kianenigma) [inline frame/balances/src/lib.rs:979]: I am not sure what is your editor's configured max-width but it is surely too low :D 

For substrate, we use 100, and the CI will rejct more than 120. Please use 100.

**Comment by kianenigma** [comment]
> If the tests are passing, I guess it works.. just the logic is very confusing... but not sure what else can be done other than micro optimizing the code flow

Wouldn't be possible to rewrite transfer to avoid nested mutation closures? That should also fix it.

**Comment by kianenigma** [comment]
This is what I would do. Some tests fail because of the except that I added but there is description of why and how to prevent. 

```diff
-               Self::try_mutate_account(dest, |to_account, _| -> DispatchResult {
-                       Self::try_mutate_account(transactor, |from_account, _| -> DispatchResult {
-                               from_account.free = from_account.free.checked_sub(&value)
-                                       .ok_or(Error::<T, I>::InsufficientBalance)?;
+               // update only transactor.
+               Self::try_mutate_account(transactor, |from_account, _| -> DispatchResult {
+                       from_account.free = from_account.free.checked_sub(&value)
+                               .ok_or(Error::<T, I>::InsufficientBalance)?;

-                               // NOTE: total stake being stored in the same type means that this could never overflow
-                               // but better to be safe than sorry.
-                               to_account.free = to_account.free.checked_add(&value).ok_or(Error::<T, I>::Overflow)?;
+                       Self::ensure_can_withdraw(
+                               transactor,
+                               value,
+                               WithdrawReasons::TRANSFER,
+                               from_account.free,
+                       )?;

-                               let ed = T::ExistentialDeposit::get();
-                               ensure!(to_account.total() >= ed, Error::<T, I>::ExistentialDeposit);
+                       let ed = T::ExistentialDeposit::get();
+                       let allow_death = existence_requirement == ExistenceRequirement::AllowDeath;
+                       let allow_death = allow_death && system::Module::<T>::allow_death(transactor);
+                       ensure!(allow_death || from_account.free >= ed, Error::<T, I>::KeepAlive);
+
+                       Ok(())
+               })?;

-                               Self::ensure_can_withdraw(
-                                       transactor,
-                                       value,
-                                       WithdrawReasons::TRANSFER,
-                                       from_account.free,
-                               )?;
+               Self::try_mutate_account(dest, |to_account, _| -> DispatchResult {
+                       // NOTE: total stake being stored in the same type means that this could never overflow
+                       // but better to be safe than sorry.
+                       to_account.free = to_account.free.checked_add(&value).ok_or(Error::<T, I>::Overflow)?;

-                               let allow_death = existence_requirement == ExistenceRequirement::AllowDeath;
-                               let allow_death = allow_death && system::Module::<T>::allow_death(transactor);
-                               ensure!(allow_death || from_account.free >= ed, Error::<T, I>::KeepAlive);
+                       let ed = T::ExistentialDeposit::get();
+                       ensure!(to_account.total() >= ed, Error::<T, I>::ExistentialDeposit);

-                               Ok(())
-                       })
-               })?;
+                       Ok(())
+               }).expect("We must ensure that we will never bail out here; needs reading to_account twice, but I think it is fine because overaly.");

                // Emit transfer event.
                Self::deposit_event(RawEvent::Transfer(transactor.clone(), dest.clone(), value));

```

These mutation closures are still a bit alien to me tbh though, maybe I am missing something.

**Comment by shamb0** [comment]
Thanks @kianenigma, for comments and suggestion to avoid nested mutation closures, will work on suggested flow and get back.

**Comment by kianenigma** [comment]
I would take my proposal with a grain of salt though, as I said I think there was an intention in nesting the mutations and probably it was important.

**Comment by shamb0** [comment]
Hello @kianenigma,  completed "fn transfer()"  implementation by avoiding nested mutation closure,
Plz find draft version in below link ...

https://github.com/paritytech/substrate/pull/7880

As you anticipated, "tests_reentrancy::has_reentrancy_issue" is passed but below test cases are failed

```shell
failures:
    tests_composite::balance_transfer_when_reserved_should_not_work
    tests_composite::transfer_keep_alive_works
    tests_local::balance_transfer_when_reserved_should_not_work
    tests_local::transfer_keep_alive_works
```

Plz find complete log here ...

https://gist.github.com/shamb0/ebc7af6b8b4b830a4074c08e1b51c425

Analysis is in progress, to find cause for this failure. In between plz let me know is there any deviation in sequence flow in draft version of the code.

**Comment by shamb0** [comment]
Hello @kianenigma,
 

Merged all changes here for easy tracking, Now all tests are passed, but with below magical work around.



I hope it's in complete state for next level of review.

**Comment by kianenigma** [comment]
Sorry for my delay, I just wanted to note that I will come back to this someday, but given the low priority, I have been postponing it a little 😄

**Comment by gavofyork** [comment]
Please check #8087 for a possible cleaner fix for this.

**Comment by shamb0** [comment]
Thanks @kianenigma & @gavofyork for the pointer, will check and get back soon.

**Comment by kianenigma** [comment]
I'm a bit confused now: #8087 fixes the situation of a `on_unbalance` inside `try_mutate` 1 level deep. Two next steps are needed: 

1. Just add a single test for that and verify it, which I think you should just make a PR on top of #8087 
2. Refactor other functions to prevent nested `try_mutates`.

Which one is this PR now?

**Comment by thiolliere** [comment]
I think this PR is also solved by https://github.com/paritytech/substrate/pull/8087
But the tests are good, IMO we can extract the test and rebase this PR on top of https://github.com/paritytech/substrate/pull/8087

File: frame/balances/src/tests_reentrancy.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		self.existential_deposit = existential_deposit;
+		self
+	}
+	pub fn monied(mut self, monied: bool) -> Self {
+		self.monied = monied;
+		if self.existential_deposit == 0 {
+			self.existential_deposit = 1;
+		}
+		self
+	}
```
Comment (thiolliere) [inline frame/balances/src/tests_reentrancy.rs:138]: you never ues `monied` I think you can remove it

**Comment by thiolliere** [comment]
superseeded by https://github.com/paritytech/substrate/pull/8129 and https://github.com/paritytech/substrate/pull/8087

**Pull request closed without merging**